### PR TITLE
OPGRU and PGRU PR [exclude special fast component]

### PR DESCRIPTION
**Pull request by GaofengCheng**
related to #1799
the reference could be added later

### CODE REVIEW

File: egs/ami/s5b/local/chain/tuning/run_tdnn_pgru_1a.sh [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+else
+  gmm_dir=exp/${mic}/$gmm
+  ali_dir=exp/${mic}/${gmm}_ali_${train_set}_sp_comb
+  lores_train_data_dir=data/$mic/${train_set}_sp_comb
+  tree_dir=exp/$mic/chain${nnet3_affix}/tree_bi${tree_affix}
+  lat_dir=exp/$mic/chain${nnet3_affix}/${gmm}_${train_set}_sp_comb_lats
+  dir=exp/$mic/chain${nnet3_affix}/tdnn_pgru${tgru_affix}_sp_bi
+fi
+
+if [ $label_delay -gt 0 ]; then dir=${dir}_ld$label_delay; fi
```
Comment (danpovey) [inline egs/ami/s5b/local/chain/tuning/run_tdnn_pgru_1a.sh]: I know this was a pre-existing issue in the scripts, but I want to address this starting from now:
These directory names are too decorated.  Can you please remove the _ld5 and the _bi from these directory names for these new scripts?  let's try to do the same for all new AMI scripts.

File: egs/fisher_swbd/s5/local/chain/run_tdnn_opgru_1a.sh [hunk 2]
```diff
@@ -0,0 +1,269 @@
+#!/bin/bash
+# ./local/chain/compare_wer_general.sh tdnn_opgru_1a_sp
+# System                tdnn_opgru_1a_sp
```
Comment (danpovey) [inline egs/fisher_swbd/s5/local/chain/run_tdnn_opgru_1a.sh]: there is no comparison here with other topologies.  Can't you at least fake it based on the info that's checked into other scripts?
Please reorganize this directory to create a tuning directory, move the scripts into tuning/, and make soft links.

File: egs/fisher_swbd/s5/local/chain/run_tdnn_opgru_1a.sh [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+# nnet3 setup, and you can skip them by setting "--stage 8" if you have already
+# run those things.
+
+suffix=
+if [ "$speed_perturb" == "true" ]; then
+  suffix=_sp
+fi
+
+dir=$dir${affix:+_$affix}
+if [ $label_delay -gt 0 ]; then dir=${dir}_ld$label_delay; fi
```
Comment (danpovey) [inline egs/fisher_swbd/s5/local/chain/run_tdnn_opgru_1a.sh]: Please remove this _ld$label_delay decoration.  
In fact, I'm a little confused about where you got this sript because a lot of aspects of it seem to not be the same as other scripts in the fisher_swbd directory.
The lang directory is differently named and you are using a different size tree.
I think something went wrong here.  
Of course, if there is a good reason for these changes, or the results are better, I'll think about that.

File: egs/swbd/s5c/local/chain/tuning/run_tdnn_opgru_1a.sh [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+# nnet3 setup, and you can skip them by setting "--stage 8" if you have already
+# run those things.
+
+suffix=
+if [ "$speed_perturb" == "true" ]; then
+  suffix=_sp
+fi
+
+dir=$dir${affix:+_$affix}
+if [ $label_delay -gt 0 ]; then dir=${dir}_ld$label_delay; fi
```
Comment (danpovey) [inline egs/swbd/s5c/local/chain/tuning/run_tdnn_opgru_1a.sh]: Again the label_delay affix.

Comment (GaofengCheng) [reply egs/swbd/s5c/local/chain/tuning/run_tdnn_opgru_1a.sh]: ok will do

File: egs/fisher_swbd/s5/local/chain/run_tdnn_opgru_1a.sh (hunk 2 above)
Comment (GaofengCheng) [reply egs/fisher_swbd/s5/local/chain/run_tdnn_opgru_1a.sh]: Ok will do

File: egs/fisher_swbd/s5/local/chain/run_tdnn_opgru_1a.sh (hunk 3 above)
Comment (GaofengCheng) [reply egs/fisher_swbd/s5/local/chain/run_tdnn_opgru_1a.sh]: The experiments are OK under Fisher+Switchboard, all model use same tree "tri6_tree". The misguiding information in the TDNN-OPGRU script is because I used switchboard scripts directly (same for TDNN-PGRU, TDNN-LSTMP). But during training, I just used model training part. This will not influence results.

I will clean up the wrong scripts on Fisher+SWBD and re-upload.

File: egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py [hunk 5]
```diff
@@ -0,0 +1,642 @@
...
+        configs.append("component name={0}.h type=TanhComponent dim={1} {2}".format(name, cell_dim, repair_nonlin_str))
+
+        configs.append("# Defining the components for other cell computations")
+        configs.append("component name={0}.h1 type=ElementwiseProductComponent input-dim={1} output-dim={2}".format(name, 2 * cell_dim, cell_dim))
+        configs.append("component name={0}.y1 type=ElementwiseProductComponent input-dim={1} output-dim={2}".format(name, 2 * cell_dim, cell_dim))
+        configs.append("component name={0}.y2 type=ElementwiseProductComponent input-dim={1} output-dim={2}".format(name, 2 * cell_dim, cell_dim))
+        configs.append("component name={0}.y type=NoOpComponent dim={1}".format(name, cell_dim))
+
+        configs.append("# Defining fixed scale/bias component for (1 - z_t)")
+        configs.append("component name={0}.fixed_scale_minus_one type=FixedScaleComponent scales={1}".format(name, self.config['vars_path']+"/minus_one"))
```
Comment (danpovey) [inline egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py]: there is no need for 'vars_path' here.  FixedScaleComponent defaults to a scale of 1.0, you just have to supply the dim; and FixedBiasComponent can have a constant bias specified by "bias=1.0 dim=xxx".

As soon as I merge this to the RNNLM branch, though, I'll ask you to simplify these config-generation scripts to use the offset and scale capabilities I added there.

File: egs/ami/s5b/local/chain/tuning/run_tdnn_pgru_1a.sh [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+  fixed-affine-layer name=lda input=Append(-1,0,1,ReplaceIndex(ivector, t, 0)) affine-transform-file=$dir/configs/lda.mat
+
+  # the first splicing is moved before the lda layer, so no splicing here
+  relu-renorm-layer name=tdnn1 dim=1024
+  relu-renorm-layer name=tdnn2 input=Append(-1,0,1) dim=1024
+  relu-renorm-layer name=tdnn3 input=Append(-1,0,1) dim=1024
+
+  # check steps/libs/nnet3/xconfig/gru.py for the other options and defaults
+  pgru-layer name=pgru1 cell-dim=1024 recurrent-projection-dim=256 non-recurrent-projection-dim=256 delay=-3 vars_path="$dir/configs"
+  relu-renorm-layer name=tdnn4 input=Append(-3,0,3) dim=1024
```
Comment (danpovey) [inline egs/ami/s5b/local/chain/tuning/run_tdnn_pgru_1a.sh]: I notice these scripts are using relu-renorm-layer.  In general we prefer relu-batchnorm-layer which tends to be better.  It would be good to have a TDNN+LSTM example checked in here too, since it is one of the baselines in your paper.  That also should have relu-batchnorm-layer.

File: egs/ami/s5b/local/chain/tuning/run_tdnn_pgru_1a.sh [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+min_seg_len=1.55
+use_ihm_ali=false
+train_set=train_cleaned
+gmm=tri3_cleaned  # the gmm for the target data
+ihm_gmm=tri3  # the gmm for the IHM system (if --use-ihm-ali true).
+num_threads_ubm=32
+nnet3_affix=_cleaned  # cleanup affix for nnet3 and chain dirs, e.g. _cleaned
+num_epochs=4
+
+chunk_width=150
```
Comment (danpovey) [inline egs/ami/s5b/local/chain/tuning/run_tdnn_pgru_1a.sh]: It would be better instead of 150, to make this 150,140,90.

File: egs/fisher_swbd/s5/local/chain/run_tdnn_opgru_1a.sh [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+  input dim=100 name=ivector
+  input dim=40 name=input
+
+  # please note that it is important to have input layer with the name=input
+  # as the layer immediately preceding the fixed-affine-layer to enable
+  # the use of short notation for the descriptor
+  fixed-affine-layer name=lda input=Append(-2,-1,0,1,2, ReplaceIndex(ivector, t, 0)) affine-transform-file=$dir/configs/lda.mat
+
+  # the first splicing is moved before the lda layer, so no splicing here
+  relu-renorm-layer name=tdnn1 dim=1024
```
Comment (danpovey) [inline egs/fisher_swbd/s5/local/chain/run_tdnn_opgru_1a.sh]: I know this is kind of not relevant to this PR, and you might want to look into it later, but I think you'd probably get improvements by adding l2 regularization to this setup.  (it's not really about regularization, it's about helping it to learn faster).  You'd probably want l2 constants that were about 10 times smaller than the ones in mini_librispeech-- maybe even smaller than that.
You would also probably get improvements in all these setups by adding non-splicing TDNN layers at the beginning after layers tdnn2 and tdnn3.  Again, not closely related but it would be great if you could try it at some point.

File: egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py [hunk 9]
```diff
@@ -0,0 +1,1150 @@
...
+
+        configs.append("# Defining the components for other cell computations")
+        configs.append("component name={0}.o1 type=ElementwiseProductComponent input-dim={1} output-dim={2}".format(name, 2 * cell_dim, cell_dim))
+        configs.append("component name={0}.y1 type=ElementwiseProductComponent input-dim={1} output-dim={2}".format(name, 2 * cell_dim, cell_dim))
+        configs.append("component name={0}.y2 type=ElementwiseProductComponent input-dim={1} output-dim={2}".format(name, 2 * cell_dim, cell_dim))
+        configs.append("component name={0}.y type=NoOpComponent dim={1}".format(name, cell_dim))
+
+        configs.append("# Defining fixed scale/bias component for (1 - z_t)")
+        configs.append("component name={0}.fixed_scale_minus_one type=FixedScaleComponent scales={1}".format(name, self.config['vars_path']+"/minus_one"))
+        configs.append("component name={0}.fixed_bias_one type=FixedBiasComponent bias={1}".format(name, self.config['vars_path']+"/bias_one"))
```
Comment (danpovey) [inline egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py]: It should not be necessary to use fixed scale and bias now that I have merged those upgrades to Descriptors.   Can you upgrade these scripts?

Also, do you have any plan to implement a fast version of the GRU comopnent?

Comment (GaofengCheng) [reply egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py]: Will it be OK to change the config into Descriptors but use the exps ran by scale-bias ?
It's time consuming to rerun all the scripts.
I'm not sure when I am OK to update a fast version OPGRU (it may need some tuning), I'm a little busy this month. I will ask Huang Lu whether he is OK.

Comment (danpovey) [reply egs/wsj/s5/steps/libs/nnet3/xconfig/gru.py]: yes that's OK, it's totally equivalent-- just run something small to make sure it doesn't crash and the objectives aren't wildly different than before for the 1st iteration.

File: egs/ami/s5b/local/chain/tuning/run_tdnn_pgru_1a.sh (hunk 1 above)
Comment (GaofengCheng) [reply egs/ami/s5b/local/chain/tuning/run_tdnn_pgru_1a.sh]: @danpovey will  it be OK to remain only 'sp' but deprecate 'ld5' and 'bi' in dir naming ?

Comment (danpovey) [reply egs/ami/s5b/local/chain/tuning/run_tdnn_pgru_1a.sh]: Yes, you can keep _sp, but remove _ld5 and _bi.

**Comment by GaofengCheng** [comment]
@ngoel17 search my latest commit under egs/fisher+swbd for tdnn_opgru. The results is showed. I have not double check the scripts, if there exist some small bugs in scripts, I believe it's not difficult to solve by yourself.

**Comment by GaofengCheng** [comment]
Comparison between descriptor  and fix-scale-bias ( swbd ):
The difference between fix-scale-bias and descriptor is very small
@danpovey 
left: fix-scale-bias; right: descriptor
1.TDNN-PGRU：
```
# ./local/chain/compare_wer_general.sh tdnn_gru_1a_like_1e_adding_initial0_ld5_sp tdnn_gru_1a_like_1e_adding_initial0_descriptor_ld5_sp
# System                tdnn_gru_1a_like_1e_adding_initial0_ld5_sp tdnn_gru_1a_like_1e_adding_initial0_descriptor_ld5_sp
# WER on train_dev(tg)      12.70     12.61
# WER on train_dev(fg)      11.77     11.61
# WER on eval2000(tg)        14.9      14.8
# WER on eval2000(fg)        13.4      13.4
# Final train prob         -0.077    -0.075
# Final valid prob         -0.092    -0.093
# Final train prob (xent)        -0.929    -0.918
# Final valid prob (xent)       -0.9934   -0.9905
```

2.TDNN-NormPGRU:
```
# ./local/chain/compare_wer_general.sh tdnn_gru_1a_like_1e_adding_initial0_batchnorm_vertical_renorm_ld5_sp tdnn_gru_1a_like_1e_adding_initial0_batchnorm_vertical_renorm_descriptor_ld5_sp
# System                tdnn_gru_1a_like_1e_adding_initial0_batchnorm_vertical_renorm_ld5_sp tdnn_gru_1a_like_1e_adding_initial0_batchnorm_vertical_renorm_descriptor_ld5_sp
# WER on train_dev(tg)      12.75     12.92
# WER on train_dev(fg)      11.90     11.95
# WER on eval2000(tg)        15.1      14.9
# WER on eval2000(fg)        13.5      13.5
# Final train prob         -0.062    -0.062
# Final valid prob         -0.082    -0.083
# Final train prob (xent)        -0.837    -0.830
# Final valid prob (xent)       -0.9326   -0.9351
```

3. TDNN-OPGRU:
```
# ./local/chain/compare_wer_general.sh tdnn_gru_1a_like_1e_adding_initial0_ogate4_notrunc_ld5_sp tdnn_gru_1a_like_1e_adding_initial0_ogate4_notrunc_descriptor_ld5_sp
# System                tdnn_gru_1a_like_1e_adding_initial0_ogate4_notrunc_ld5_sp tdnn_gru_1a_like_1e_adding_initial0_ogate4_notrunc_descriptor_ld5_sp
# WER on train_dev(tg)      12.61     12.62
# WER on train_dev(fg)      11.55     11.65
# WER on eval2000(tg)        14.9      14.9
# WER on eval2000(fg)        13.4      13.6
# Final train prob         -0.065    -0.067
# Final valid prob         -0.086    -0.087
# Final train prob (xent)        -0.871    -0.877
# Final valid prob (xent)       -0.9628   -0.9783
```

4. BatchTDNN-NormOPGRU (recommended)
```
# ./local/chain/compare_wer_general.sh --looped tdnn_gru_1a_like_1e_adding_initial0_ogate4_notrunc_batch_renorm_tdnnbatch_perframe0.2_ld5_sp tdnn_gru_1a_like_1e_adding_initial0_ogate4_notrunc_batch_renorm_tdnnbatch_perframe0.2_descriptor_ld5_sp
# System                tdnn_gru_1a_like_1e_adding_initial0_ogate4_notrunc_batch_renorm_tdnnbatch_perframe0.2_ld5_sp tdnn_gru_1a_like_1e_adding_initial0_ogate4_notrunc_batch_renorm_tdnnbatch_perframe0.2_descriptor_ld5_sp
# WER on train_dev(tg)      12.39     12.29
#           [looped:]       12.32     12.27
# WER on train_dev(fg)      11.39     11.32
#           [looped:]       11.35     11.34
# WER on eval2000(tg)        15.1      14.8
#           [looped:]        15.1      14.9
# WER on eval2000(fg)        13.6      13.5
#           [looped:]        13.5      13.5
# Final train prob         -0.066    -0.065
# Final valid prob         -0.085    -0.084
# Final train prob (xent)        -0.889    -0.888
# Final valid prob (xent)       -0.9837   -0.9824
```

**Comment by danpovey** [comment]
OK, I think that's just the noise from running it again.

File: egs/ami/s5b/local/chain/tuning/run_tdnn_opgru_1c.sh [hunk 10]
```diff
@@ -0,0 +1,349 @@
...
+    --chain.apply-deriv-weights false \
+    --chain.lm-opts="--num-extra-lm-states=2000" \
+    --egs.dir "$common_egs_dir" \
+    --egs.opts "--frames-overlap-per-eg 0" \
+    --egs.chunk-width $chunk_width \
+    --egs.chunk-left-context $chunk_left_context \
+    --egs.chunk-right-context $chunk_right_context \
+    --trainer.dropout-schedule $dropout_schedule \
+    --trainer.optimization.backstitch-training-scale 1 \
+    --trainer.optimization.backstitch-training-interval 4 \
```
Comment (whsunset) [inline egs/ami/s5b/local/chain/tuning/run_tdnn_opgru_1c.sh]: Should it be scale=0.3 and interval=1, as said in the comment above?

Comment (GaofengCheng) [reply egs/ami/s5b/local/chain/tuning/run_tdnn_opgru_1c.sh]: yes, I will make a fix commit.
